### PR TITLE
Add support of raw_body hooks

### PR DESCRIPTION
**Pull request by socketpair**
This simplifies writing of transparent HTTP-proxies with this library

### CODE REVIEW

**Comment by socketpair** [comment]
Unfortunatelly I don't know how to add tests...

File: htp/htp_config_private.h [hunk 1]
```diff
@@ -238,6 +238,12 @@ struct htp_cfg_t {
     htp_hook_t *hook_request_body_data;
 
     /**
+     * Receives raw request body data, starting immediatelly after empty line
+     * that terminates list of request headers.
+     */
+    htp_hook_t *hook_request_raw_body_data;
```
Comment (inliniac) [inline htp/htp_config_private.h]: To maintain binary compatibility this should be added to the end of the structure.

File: htp/htp_config_private.h [hunk 2]
```diff
@@ -296,6 +302,13 @@ struct htp_cfg_t {
     htp_hook_t *hook_response_body_data;
 
     /**
+     * Receives raw response body data, starting immediatelly after empty line
+     * that terminates list of response headers.
+     */
+    htp_hook_t *hook_response_raw_body_data;
```
Comment (inliniac) [inline htp/htp_config_private.h]: Same here.

**Comment by inliniac** [comment]
You can take this commit as an example for adding tests https://github.com/OISF/libhtp/commit/b85cb3a24369b5c6dbef74da1307edd32a64b403

File: htp/htp_config_private.h (hunk 1 above)
Comment (socketpair) [inline htp/htp_config_private.h]: This is not user-accessible structure. Should I fix that? So, if user upgrades libhtp, binary compatibility issue should never happen. Does not it ?

File: test/test_main.cpp [hunk 3]
```diff
@@ -1914,3 +1914,50 @@ TEST_F(ConnectionParsing, RequestUriTooLarge) {
...
+
+    ASSERT_EQ(1, htp_list_size(connp->conn->transactions));
+
+    htp_tx_t *tx = (htp_tx_t *) htp_list_get(connp->conn->transactions, 0);
+    ASSERT_TRUE(tx != NULL);
+
+    ASSERT_EQ(HTP_REQUEST_COMPLETE, tx->request_progress);
+    ASSERT_EQ(HTP_RESPONSE_COMPLETE, tx->response_progress);
+
+    printf("req: %s\n", collected_req.c_str());
```
Comment (socketpair) [inline test/test_main.cpp]: these `printf`s are not suited for complete code, but for intermediate debugging only.

Callbacks are not fired. What I'm doing wrong?

File: test/test_main.cpp [hunk 4]
```diff
@@ -1914,3 +1914,50 @@ TEST_F(ConnectionParsing, RequestUriTooLarge) {
...
+
+
+static std::string collected_req;
+static int collect_raw_reqest(htp_tx_data_t *d) {
+    printf("reqc: %d\n", (int)d->len);
+    collected_req += std::string((const char*)d->data, d->len);
+    return HTP_OK;
+}
+
+static std::string collected_resp;
```
Comment (socketpair) [inline test/test_main.cpp:1926]: Should I replace it with `bstr` ?

File: htp/htp_config_private.h (hunk 1 above)
Comment (inliniac) [inline htp/htp_config_private.h]: Even thought it's in a file named 'private' it actually exposed. At least Suricata uses it, and even though we never seem to access the structure directly, I am rather safe that sorry. When we bump to 0.6.x we can reshuffle it again. A similar case is here: https://github.com/OISF/libhtp/commit/5b43affd3c9bc1c4280a54266a92f6d0ad7600d1#diff-9815a4e5141bd3d3914c1b5f9295fdecR337

File: test/test_main.cpp (hunk 3 above)
Comment (inliniac) [inline test/test_main.cpp]: Where in your patch are you invoking the new callbacks?

Comment (socketpair) [inline test/test_main.cpp]: nowhere. Due to changes in `htp/htp_request.c` and `htp/htp_response.c` it should be called automatically. does not it? I'm novice in libhtp, and make patch as copy-paste with `headers_data` callbacks. Maybe I forgot something ?

File: htp/htp_config_private.h (hunk 1 above)
Comment (socketpair) [inline htp/htp_config_private.h]: fixed

File: htp/htp_config_private.h (hunk 2 above)
Comment (socketpair) [inline htp/htp_config_private.h]: fixed

File: test/test_main.cpp [hunk 5]
```diff
@@ -1914,3 +1914,44 @@ TEST_F(ConnectionParsing, RequestUriTooLarge) {
...
+    ASSERT_EQ(1, htp_list_size(connp->conn->transactions));
+
+    htp_tx_t *tx = (htp_tx_t *) htp_list_get(connp->conn->transactions, 0);
+    ASSERT_TRUE(tx != NULL);
+
+    ASSERT_EQ(HTP_REQUEST_COMPLETE, tx->request_progress);
+    ASSERT_EQ(HTP_RESPONSE_COMPLETE, tx->response_progress);
+
+    ASSERT_EQ("3\r\nqwe\r\n3\r\nasd\r\n0\r\n", collected_req);
+    ASSERT_EQ("3\r\nxxx\r\n3\r\nyyy\r\n0\r\n", collected_resp);
```
Comment (socketpair) [inline test/test_main.cpp]: req test is passed! yay! BUT!
resp test failed due to:

```
test_main.cpp:1953: Failure
Value of: collected_resp
  Actual: "3\r\nxxx\n3\r\nyyy\r\n0\r\n"
Expected: "3\r\nxxx\r\n3\r\nyyy\r\n0\r\n"
```

as you can see libhtp strip `\r` somewhere!

Comment (socketpair) [inline test/test_main.cpp:1952]: @inliniac sorry. I have fixed bugs in my tests. Now everything is fine.

**Comment by socketpair** [comment]
Well, @inliniac, why not to merge ?

**Comment by socketpair** [comment]
I'm not interested in libhtp anymore

**Pull request closed without merging**